bootloader/lk.go

```diff
@@ -0,0 +1,211 @@
+		return filepath.Join(dirs.GlobalRootDir, "/boot/lk/")
+	}
+}
+
+func (l *lk) ConfigFile() string {
+	return l.envFile()
+}
+
+func (l *lk) envFile() string {
+	// as for dir, we have two scenarios, image building and runtime
+	if inRuntimeMode() {
+		// TO-DO: this should be eventually fetched from gadget.yaml
+		return filepath.Join(l.dir(), "snapbootsel")
+	} else {
+		return filepath.Join(l.dir(), "snapbootsel.bin")
+	}
+}
+
+// determine mode we are in, runtime or image build
+func inRuntimeMode() bool {
+	if dirs.GlobalRootDir == "/" {
+		return true
+	} else {
+		return false
+	}
+}
+
+func (l *lk) GetBootVars(names ...string) (map[string]string, error) {
+	out := make(map[string]string)
+
+	env := lkenv.NewEnv(l.envFile())
+	if err := env.Load(); err != nil {
+		return nil, err
+	}
+
+	for _, name := range names {
+		out[name] = env.Get(name)
+	}
+
+	return out, nil
+}
+
+func (l *lk) SetBootVars(values map[string]string) error {
+	env := lkenv.NewEnv(l.envFile())
+	if err := env.Load(); err != nil && !os.IsNotExist(err) {
+		return err
+	}
+
+	// update environment only if something change
+	dirty := false
+	for k, v := range values {
+		// already set to the right value, nothing to do
+		if env.Get(k) == v {
+			continue
+		}
+		env.Set(k, v)
+		dirty = true
+	}
+
+	if dirty {
+		return env.Save()
+	}
+
+	return nil
+}
+
+// for lk we need to flash boot image to free bootimg partition
+// first make sure there is free boot part to use
+// if this is image creation, we just extract file
+func (l *lk) ExtractKernelAssets(s snap.PlaceInfo, snapf snap.Container) error {
+	blobName := filepath.Base(s.MountFile())
+
+	logger.Debugf("ExtractKernelAssets (%s)\n", blobName)
+
+	env := lkenv.NewEnv(l.envFile())
+	if err := env.Load(); err != nil && !os.IsNotExist(err) {
+		return err
+	}
+
+	bootPartition, err := env.FindFreeBootPartition(blobName)
+	if err != nil {
+		return err
+	}
+
+	if inRuntimeMode() {
+		logger.Debugf("ExtractKernelAssets handling run time usecase\n")
+		// this is live system, extracted bootimg needs to be flashed to
+		// free bootimg partition and env has be updated boot slop mapping
+		tmpdir, err := ioutil.TempDir("", "bootimg")
+		if err != nil {
+			return fmt.Errorf("Failed to create tmp directory %v", err)
+		}
+		defer os.RemoveAll(tmpdir)
+		if err := snapf.Unpack(env.GetBootImageName(), tmpdir); err != nil {
+			return fmt.Errorf("Failed to unpack %s %v", env.GetBootImageName(), err)
+		}
+		// write boot.img to free boot partition
+		bootimgName := filepath.Join(tmpdir, env.GetBootImageName())
+		bif, err := os.Open(bootimgName)
+		if err != nil {
+			return fmt.Errorf("Failed to open unpacked %s %v", env.GetBootImageName(), err)
+		}
+		defer bif.Close()
+		bpart := filepath.Join(l.dir(), bootPartition)
+
+		bpf, err := os.OpenFile(bpart, os.O_WRONLY, 0660)
+		if err != nil {
+			return fmt.Errorf("Failed to open boot partition [%s] %v", bpart, err)
+		}
+		defer bpf.Close()
+
+		buf := make([]byte, 1024)
```

Comment:
Silly question - could we use "io.Copy" here? Something like:
```
diff --git a/bootloader/lk.go b/bootloader/lk.go
index 5b554c702f..cdaef2477b 100644
--- a/bootloader/lk.go
+++ b/bootloader/lk.go
@@ -167,20 +168,8 @@ func (l *lk) ExtractKernelAssets(s snap.PlaceInfo, snapf snap.Container) error {
                }
                defer bpf.Close()
 
-               buf := make([]byte, 1024)
-               for {
-                       // read by chunks
-                       n, err := bif.Read(buf)
-                       if err != nil && err != io.EOF {
-                               return fmt.Errorf("Failed to read buffer chunk of %s %v", env.GetBootImageName(), err)
-                       }
-                       if n == 0 {
-                               break
-                       }
-                       // write a chunk
-                       if _, err := bpf.Write(buf[:n]); err != nil {
-                               return fmt.Errorf("Failed to write buffer chunk of %s %v", env.GetBootImageName(), err)
-                       }
+               if _, err := io.Copy(bpf, bif); err != nil {
+                       return err
                }
        } else {
                // we are preparing image, just extract boot image to bootloader directory

```
Doing this seems to not break anything - so either we don't have tests for this or it works :)

Reply:
I'm trying to remember reason I did it by chunks, but indeed if `io.Copy` works, then we should use it instead